### ISSUE TITLE
Entity/DB has a null bounding-box error loading INTEGER .PCD file

### ISSUE THREAD

**Issue by chris2b5** [CloudCompare/CloudCompare#1433]
**Describe the bug**

When reading a .pcd file (ascii or binary) as float data type (.pcd TYPE field set to F) everything is fine, but reading .pcd files with integer data type (.pcd TYPE field set to I) the error `[ccGLWindow] Entity/DB has a null bounding-box!` is reported in the console and the pointcloud is not displayed.

**How to reproduce**

**Minimal reproducible example:**

Save the below to a text file with .pcd extension. Use File -> Load dialog to load the file. The point cloud will be displayed properly (it is adapted from this [example]( https://pcl.readthedocs.io/projects/tutorials/en/latest/pcd_file_format.html#example) )
```
# .PCD v.7 - Point Cloud Data file format
VERSION .7
FIELDS x y z
SIZE 4 4 4
TYPE F F F
COUNT 1 1 1
WIDTH 213
HEIGHT 1
VIEWPOINT 0 0 0 1 0 0 0
POINTS 213
DATA ascii
937 337 0 
908 356 0 
819 32 0 
971 278 0 
944 294 0 
981 242 0 
936 261 0 
916 274 0 
819 293 0 
907 241 0 
832 233 0 
991 211 0 
892 211 0 
850 212 0 
810 322 0 
744 321 0 
699 322 0 
810 293 0 
755 297 0 
810 243 0 
749 247 0 
680 297 0 
665 290 0 
694 252 0 
628 221 0 
587 321 0 
521 319 0 
493 322 0 
443 321 0 
586 292 0 
534 291 0 
593 241 0 
535 244 0 
500 312 0 
441 297 0 
507 221 0 
439 239 0 
810 211 0 
735 211 0 
699 210 0 
633 211 0 
591 212 0 
524 214 0 
493 210 0 
438 196 0 
434 160 0 
378 321 0 
334 321 0 
238 321 0 
378 293 0 
330 310 0 
378 243 0 
302 291 0 
234 294 0 
294 243 0 
235 241 0 
188 322 0 
159 321 0 
086 321 0 
039 322 0 
200 312 0 
141 295 0 
209 223 0 
138 242 0 
085 294 0 
048 312 0 
086 243 0 
378 211 0 
294 193 0 
237 193 0 
294 160 0 
237 160 0 
190 210 0 
157 212 0 
086 211 0 
012 321 0 
-075 320 0 
-106 323 0 
-163 321 0 
000 294 0 
-057 294 0 
-003 243 0 
-055 241 0 
-109 313 0 
-150 294 0 
-115 220 0 
-151 243 0 
-191 321 0 
-281 318 0 
-307 323 0 
-338 321 0 
-212 298 0 
-263 299 0 
-225 243 0 
-250 237 0 
-299 311 0 
-358 295 0 
-292 222 0 
-361 241 0 
-003 211 0 
-073 213 0 
-105 209 0 
-136 214 0 
-225 193 0 
-244 197 0 
-225 160 0 
-234 160 0 
-307 209 0 
-341 212 0 
-405 321 0 
-505 321 0 
-417 308 0 
-442 288 0 
-415 220 0 
-448 242 0 
-506 293 0 
-509 242 0 
-570 223 0 
-596 321 0 
-651 321 0 
-725 321 0 
-755 323 0 
-596 293 0 
-650 293 0 
-594 242 0 
-650 243 0 
-706 295 0 
-762 312 0 
-703 241 0 
-770 221 0 
-415 210 0 
-450 198 0 
-447 167 0 
-520 212 0 
-551 210 0 
-596 211 0 
-650 211 0 
-722 212 0 
-755 209 0 
-783 319 0 
-875 321 0 
-905 322 0 
-955 341 0 
-797 292 0 
-856 294 0 
-799 243 0 
-852 242 0 
-911 312 0 
-950 295 0 
-922 221 0 
-949 243 0 
-786 215 0 
-870 212 0 
-906 209 0 
-937 214 0 
222 -029 0 
270 -078 0 
244 -056 0 
273 -106 0 
261 -106 0 
270 -134 0 
244 -155 0 
190 -025 0 
142 -024 0 
198 -037 0 
142 -035 0 
093 -024 0 
044 -024 0 
093 -035 0 
044 -035 0 
211 -173 0 
029 -125 0 
036 -156 0 
224 -205 0 
220 -236 0 
212 -208 0 
193 -256 0 
144 -274 0 
030 -187 0 
102 -274 0 
168 -283 0 
138 -286 0 
086 -295 0 
044 -315 0 
-006 -024 0 
-057 -024 0 
-006 -035 0 
-057 -035 0 
-106 -024 0 
-152 -025 0 
-106 -035 0 
-160 -037 0 
008 -125 0 
000 -156 0 
-174 -173 0 
-185 -029 0 
-207 -056 0 
-233 -077 0 
-224 -106 0 
-236 -106 0 
-207 -156 0 
-233 -135 0 
006 -187 0 
-066 -274 0 
-175 -207 0 
-108 -274 0 
-155 -257 0 
-007 -315 0 
-050 -295 0 
-103 -286 0 
-131 -283 0 
-187 -205 0 
-183 -237 0 
```

Now restart CloudCompare (or just delete the data in the DB Tree) and change the line `TYPE F F F` to `TYPE I I I` in the file, and repeat the process. This time the pointcloud will not be displayed and the error will be reported.

**Expected behaviour**

The integer version of the pointcloud file should be read in just the same way as the float version.

**Additional context**

The real issue I am faxing is with *binary* .pcd files with 16-bit integer xyz data. However I hope the above *ascii* version is reproducing the same problem that I am seeing with the binary files...

**Your environment**
 - CC Version (or git hash/tag): v2.12 alpha
 - OS & Version: Win7 64-bit
 - Graphics card: NVIDIA Quadro M620
 - Qt Version (if compiling): not compiling, used windows installer.

**Comment by dgirardeau** [comment]
I looked at the code, and it seems PCL fails to load the PCD file because CloudCompare requests it to read a 'pcl::PointXYZ' cloud. Here 'pcl::PointXYZ' represents a 3D point with floating point coordinates.

And PCL refuses to load integer coordinates in a pcl::PointXYZ point because the type doesn't match 😓 (due to the way it works internally). And I also see PCL doesn't provide any 'integer' point.

So I had to create a custom type. At least it seems to work...

We will have to wait for this PR to be merged: https://github.com/CloudCompare/CloudCompare/pull/1435

**Comment by dgirardeau** [comment]
Ok the PR is merged. And I updated the 2.12.alpha version (for Windows) by the way.

I'll let you test.

**Comment by chris2b5** [comment]
Thanks very much for a very quick response!

I downloaded and installed the new alpha and did some testing.

The good news is that integer formats (both binary and ascii) are working at 'size' 4 i.e. 4 bytes per value.
However the bad news is that integer formats still show the same error behaviour at 'size' 2 i.e. 2 bytes per value.

I have attached the files I created to do the testing. The type_F and type_I_size_4 files all work fine. The type_I_size_2 files don't work.

[pcd_test_files.zip](https://github.com/CloudCompare/CloudCompare/files/6222126/pcd_test_files.zip)

**Comment by dgirardeau** [comment]
Yes that's due to how PCL manages its points clouds/files. Clearly integer coordinates are not standard. And adding another specific case for 'short' coordinates really looks... silly 😅

**Comment by WargodHernandez** [comment]
To be fair the file format specifies support for sizes 1,2,4, & 8 and also types I, U, & F. It doesn't specifically say which type and size are applicable to which field. 

No matter what these types will all just be casted to float for us in CC, or double if it was compiled with that switch.

https://pointclouds.org/documentation/tutorials/pcd_file_format.html

![Screenshot_20210329-080951~2](https://user-images.githubusercontent.com/7495752/112858018-50b47600-9066-11eb-8fd9-8073cc477863.png)

**Comment by dgirardeau** [comment]
Yes, but they don't even provide the point types for 'I' themselves 😅

**Comment by WargodHernandez** [comment]
Yeah, that's typical for PCL as the normal PCL way to do this stuff is to use templates for your exact cloud format. The loading of a PCD with unknown cloud structure at compile time is the abnormal situation.

**Comment by dgirardeau** [comment]
On the other hand, they do provide the PointXYZ structure which is using float coordinates... But for me that's logical since it's definitely 95% of the use cases I guess?

**Comment by dgirardeau** [comment]
Ok, second attempt: https://github.com/CloudCompare/CloudCompare/pull/1439

I made some code changes so as to make the support of new point types less painful.

CC should now handle signed integers of 16 and 32bits, as well as doubles (just in case).

**Comment by dgirardeau** [comment]
And... a new version is now online! I'll let you test, and hopefully this time all the weird cases will be covered (don't speak to me about a cloud with unsigned 8bit :p

**Comment by dgirardeau** [comment]
@chris2b5 Any chance you can test the latest 2.12 alpha version?

**Comment by chris2b5** [comment]
Great!

Tested working for:

| FIELDS | TYPE | SIZE | DATA  |
|---         | ---    | ---    | ---      |
| x y z    | F        | 4      | ascii    |
| x y z    | F        | 4      | binary |
| x y z    | I         | 4      | ascii    |
| x y z    | I         | 4      | binary |
| x y z    | I         | 2      | ascii    |
| x y z    | I         | 2      | binary |

I also tried some real, larger datasets (~ 1M points), all good.

I may generate some test files for some of the weirder cases too, if I get some time over the next few days. But for now, this does everything I need it to :+1:

**Comment by WargodHernandez** [comment]
Based on the last comment this appears to be finished.

**Comment by stuartbourne** [comment]
This may not be the right place for this, but I noticed you mentioned support for doubles. Attempting to import a simple example results in a `[ccGLWindow] Entity/DB has a null bounding-box! Can't zoom in...`. Example is as follows:

```
# .PCD v0.7 - Point Cloud Data file format
VERSION 0.7
FIELDS x y z
SIZE 8 8 8
TYPE F F F
COUNT 1 1 1
WIDTH 288
HEIGHT 1
VIEWPOINT 0 0 0 1 0 0 0
POINTS 288
DATA ascii
9999999.9139 0.9503 0.3622
9999999.9129 0.9505 0.3622
9999999.9119 0.9507 0.3622
9999999.9110 0.9509 0.3622
9999999.9100 0.9511 0.3622
9999999.9090 0.9513 0.3622
9999999.9080 0.9515 0.3622
9999999.9070 0.9516 0.3622
9999999.9060 0.9518 0.3622
9999999.9051 0.9520 0.3622
9999999.9041 0.9522 0.3622
9999999.9031 0.9524 0.3622
9999999.9021 0.9526 0.3622
9999999.9011 0.9527 0.3622
9999999.9001 0.9529 0.3622
9999999.8992 0.9531 0.3622
9999999.8982 0.9533 0.3622
9999999.8972 0.9535 0.3622
9999999.8962 0.9537 0.3622
9999999.8952 0.9539 0.3622
9999999.8942 0.9540 0.3622
9999999.8933 0.9542 0.3622
9999999.8923 0.9544 0.3622
9999999.8913 0.9546 0.3622
9999999.8903 0.9548 0.3622
9999999.8893 0.9550 0.3622
9999999.8884 0.9551 0.3622
9999999.8874 0.9553 0.3622
9999999.8864 0.9555 0.3622
9999999.8854 0.9557 0.3622
9999999.8844 0.9559 0.3622
9999999.8834 0.9561 0.3622
9999999.8825 0.9563 0.3622
9999999.8815 0.9564 0.3622
9999999.8805 0.9566 0.3622
9999999.8795 0.9568 0.3622
9999999.8785 0.9570 0.3622
9999999.8775 0.9572 0.3622
9999999.8766 0.9574 0.3622
9999999.8756 0.9575 0.3622
9999999.8746 0.9577 0.3622
9999999.8736 0.9579 0.3622
9999999.8726 0.9581 0.3622
9999999.8716 0.9583 0.3622
9999999.8707 0.9585 0.3622
9999999.8697 0.9587 0.3622
9999999.8687 0.9588 0.3622
9999999.8677 0.9590 0.3622
9999999.8667 0.9592 0.3622
9999999.8657 0.9594 0.3622
9999999.8648 0.9596 0.3622
9999999.8638 0.9598 0.3622
9999999.8628 0.9599 0.3622
9999999.8618 0.9601 0.3622
9999999.8608 0.9603 0.3622
9999999.8598 0.9605 0.3622
9999999.8589 0.9607 0.3622
9999999.8579 0.9609 0.3622
9999999.8569 0.9611 0.3622
9999999.8559 0.9612 0.3622
9999999.8549 0.9614 0.3622
9999999.8540 0.9616 0.3622
9999999.8530 0.9618 0.3622
9999999.8520 0.9620 0.3622
9999999.8510 0.9622 0.3622
9999999.8500 0.9623 0.3622
9999999.8490 0.9625 0.3622
9999999.8481 0.9627 0.3622
9999999.8471 0.9629 0.3622
9999999.8461 0.9631 0.3622
9999999.8451 0.9633 0.3622
9999999.8441 0.9635 0.3622
9999999.8431 0.9636 0.3622
9999999.8422 0.9638 0.3622
9999999.8412 0.9640 0.3622
9999999.8402 0.9642 0.3622
9999999.8392 0.9644 0.3622
9999999.8382 0.9646 0.3622
9999999.8372 0.9647 0.3622
9999999.8363 0.9649 0.3622
9999999.8353 0.9651 0.3622
9999999.8343 0.9653 0.3622
9999999.8333 0.9655 0.3622
9999999.8323 0.9657 0.3622
9999999.8313 0.9659 0.3622
9999999.8304 0.9660 0.3622
9999999.8294 0.9662 0.3622
9999999.8284 0.9664 0.3622
9999999.8274 0.9666 0.3622
9999999.8264 0.9668 0.3622
9999999.8254 0.9670 0.3622
9999999.8245 0.9671 0.3622
9999999.8235 0.9673 0.3622
9999999.8225 0.9675 0.3622
9999999.8215 0.9677 0.3622
9999999.8205 0.9679 0.3622
9999999.8196 0.9681 0.3622
9999999.8186 0.9683 0.3622
9999999.8176 0.9684 0.3622
9999999.8166 0.9686 0.3622
9999999.8156 0.9688 0.3622
9999999.8146 0.9690 0.3622
9999999.8137 0.9692 0.3622
9999999.8127 0.9694 0.3622
9999999.8117 0.9695 0.3622
9999999.8107 0.9697 0.3622
9999999.8097 0.9699 0.3622
9999999.8087 0.9701 0.3622
9999999.8078 0.9703 0.3622
9999999.8068 0.9705 0.3622
9999999.8058 0.9706 0.3622
9999999.8048 0.9708 0.3622
9999999.8038 0.9710 0.3622
9999999.8028 0.9712 0.3622
9999999.8019 0.9714 0.3622
9999999.8009 0.9716 0.3622
9999999.7999 0.9718 0.3622
9999999.7989 0.9719 0.3622
9999999.7979 0.9721 0.3622
9999999.7969 0.9723 0.3622
9999999.7960 0.9725 0.3622
9999999.7950 0.9727 0.3622
9999999.7940 0.9729 0.3622
9999999.7930 0.9730 0.3622
9999999.7920 0.9732 0.3622
9999999.7911 0.9734 0.3622
9999999.7901 0.9736 0.3622
9999999.7891 0.9738 0.3622
9999999.7881 0.9740 0.3622
9999999.7871 0.9742 0.3622
9999999.7861 0.9743 0.3622
9999999.7852 0.9745 0.3622
9999999.7842 0.9747 0.3622
9999999.7832 0.9749 0.3622
9999999.7822 0.9751 0.3622
9999999.7812 0.9753 0.3622
9999999.7802 0.9754 0.3621
9999999.7793 0.9756 0.3621
9999999.7783 0.9758 0.3621
9999999.7773 0.9760 0.3621
9999999.7763 0.9762 0.3621
9999999.7753 0.9764 0.3621
9999999.7743 0.9766 0.3621
9999999.7734 0.9767 0.3621
9999999.7724 0.9769 0.3621
9999999.7714 0.9771 0.3621
9999999.7704 0.9773 0.3621
9999999.7694 0.9775 0.3621
9999999.7684 0.9777 0.3621
9999999.7675 0.9778 0.3621
9999999.7665 0.9780 0.3621
9999999.7655 0.9782 0.3621
9999999.7645 0.9784 0.3621
9999999.7635 0.9786 0.3621
9999999.7625 0.9788 0.3621
9999999.7616 0.9790 0.3621
9999999.7606 0.9791 0.3621
9999999.7596 0.9793 0.3621
9999999.7586 0.9795 0.3621
9999999.7576 0.9797 0.3621
9999999.7567 0.9799 0.3621
9999999.7557 0.9801 0.3621
9999999.7547 0.9802 0.3621
9999999.7537 0.9804 0.3621
9999999.7527 0.9806 0.3621
9999999.7517 0.9808 0.3621
9999999.7508 0.9810 0.3621
9999999.7498 0.9812 0.3621
9999999.7488 0.9814 0.3621
9999999.7478 0.9815 0.3621
9999999.7468 0.9817 0.3621
9999999.7458 0.9819 0.3621
9999999.7449 0.9821 0.3621
9999999.7439 0.9823 0.3621
9999999.7429 0.9825 0.3621
9999999.7419 0.9826 0.3621
9999999.7409 0.9828 0.3621
9999999.7399 0.9830 0.3621
9999999.7390 0.9832 0.3621
9999999.7380 0.9834 0.3621
9999999.7370 0.9836 0.3621
9999999.7360 0.9838 0.3621
9999999.7350 0.9839 0.3621
9999999.7340 0.9841 0.3621
9999999.7331 0.9843 0.3621
9999999.7321 0.9845 0.3621
9999999.7311 0.9847 0.3621
9999999.7301 0.9849 0.3621
9999999.7291 0.9850 0.3621
9999999.7282 0.9852 0.3621
9999999.7272 0.9854 0.3621
9999999.7262 0.9856 0.3621
9999999.7252 0.9858 0.3621
9999999.7242 0.9860 0.3621
9999999.7232 0.9862 0.3621
9999999.7223 0.9863 0.3621
9999999.7213 0.9865 0.3621
9999999.7203 0.9867 0.3621
9999999.7193 0.9869 0.3621
9999999.7183 0.9871 0.3621
9999999.7173 0.9873 0.3621
9999999.7164 0.9874 0.3621
9999999.7154 0.9876 0.3621
9999999.7144 0.9878 0.3621
9999999.7134 0.9880 0.3621
9999999.7124 0.9882 0.3621
9999999.7114 0.9884 0.3621
9999999.7105 0.9886 0.3621
9999999.7095 0.9887 0.3621
9999999.7085 0.9889 0.3621
9999999.7075 0.9891 0.3621
9999999.7065 0.9893 0.3621
9999999.7055 0.9895 0.3621
9999999.7046 0.9897 0.3621
9999999.7036 0.9898 0.3621
9999999.7026 0.9900 0.3621
9999999.7016 0.9902 0.3621
9999999.7006 0.9904 0.3621
9999999.6996 0.9906 0.3621
9999999.6987 0.9908 0.3621
9999999.6977 0.9910 0.3621
9999999.6967 0.9911 0.3621
9999999.6957 0.9913 0.3621
9999999.6947 0.9915 0.3621
9999999.6938 0.9917 0.3621
9999999.6928 0.9919 0.3621
9999999.6918 0.9921 0.3621
9999999.6908 0.9922 0.3621
9999999.6898 0.9924 0.3621
9999999.6888 0.9926 0.3621
9999999.6879 0.9928 0.3621
9999999.6869 0.9930 0.3621
9999999.6859 0.9932 0.3621
9999999.6849 0.9934 0.3621
9999999.6839 0.9935 0.3621
9999999.6829 0.9937 0.3621
9999999.6820 0.9939 0.3621
9999999.6810 0.9941 0.3621
9999999.6800 0.9943 0.3621
9999999.6790 0.9945 0.3621
9999999.6780 0.9946 0.3621
9999999.6770 0.9948 0.3621
9999999.6761 0.9950 0.3621
9999999.6751 0.9952 0.3621
9999999.6741 0.9954 0.3621
9999999.6731 0.9956 0.3621
9999999.6721 0.9958 0.3621
9999999.6711 0.9959 0.3621
9999999.6702 0.9961 0.3621
9999999.6692 0.9963 0.3621
9999999.6682 0.9965 0.3621
9999999.6672 0.9967 0.3621
9999999.6662 0.9969 0.3621
9999999.6652 0.9970 0.3621
9999999.6643 0.9972 0.3621
9999999.6633 0.9974 0.3621
9999999.6623 0.9976 0.3621
9999999.6613 0.9978 0.3621
9999999.6603 0.9980 0.3621
9999999.6594 0.9981 0.3621
9999999.6584 0.9983 0.3621
9999999.6574 0.9985 0.3621
9999999.6564 0.9987 0.3621
9999999.6554 0.9989 0.3621
9999999.6544 0.9991 0.3621
9999999.6535 0.9993 0.3621
9999999.6525 0.9994 0.3621
9999999.6515 0.9996 0.3621
9999999.6505 0.9998 0.3621
9999999.6495 1.0000 0.3621
9999999.6485 1.0002 0.3621
9999999.6476 1.0004 0.3621
9999999.6466 1.0005 0.3620
9999999.6456 1.0007 0.3620
9999999.6446 1.0009 0.3620
9999999.6436 1.0011 0.3620
9999999.6426 1.0013 0.3620
9999999.6417 1.0015 0.3620
9999999.6407 1.0017 0.3620
9999999.6397 1.0018 0.3620
9999999.6387 1.0020 0.3620
9999999.6377 1.0022 0.3620
9999999.6367 1.0024 0.3620
9999999.6358 1.0026 0.3620
9999999.6348 1.0028 0.3620
9999999.6338 1.0029 0.3620
9999999.6328 1.0031 0.3620
9999999.6318 1.0033 0.3620
```

**Comment by dgirardeau** [comment]
Sorry for the delay. I was able to understand the issue. I updated the code so that double values are more properly managed, and so that the Global Shift & Scale dialog can appear when using large coordinates.

I also observed an issue because the 'sensor' is very far from the points. So for now, when saving such a cloud (with a sensor attached), CC will automatically detect that it's too far from the points, and it will 'discard' the sensor and replace it by something different (derived from the Global Shift value). It's not ideal, but at least the original point coordinates and the numerical accuracy can be preserved.

I've updated the latest 2.13.beta version of CloudCompare online.
